File: docs/archive/conus-sf-integration/sdimax/localized_sdimax.py
"""Localized maximum SDI for CONUS growth-and-yield models.

Replaces the species-weighted maximum SDI (biased high, near-zero plot-level skill)
with a per-stand, data-derived maximum from an FIA-based surface. Validated to predict
observed FIA self-thinning ~85% better than species-weighting (deviance explained 0.11
vs 0.06), nationally. Model-agnostic: returns a per-stand max SDI (trees/ha) that any
G&Y engine can consume as its density limit. For FVS specifically, also emits the
per-stand SDIMAX keyword block.

Sources, in order of fidelity:
  1) TreeMap 2022 CONUS 30 m SDImax raster (Zenodo 10.5281/zenodo.19509367) -- per pixel.
  2) brms FIA plot-level SDImax table -- per FIA plot (the FIA-plot form of the same).
  3) forest-type + geography model -- portable closed-form fallback (R2 ~0.20).
"""
from __future__ import annotations
import os, math
from functools import lru_cache

TPA_PER_HA = 2.4710538  # divide trees/ha by this to get trees/acre (FVS units)

# ---------------------------------------------------------------------------
# Source 2: brms FIA plot-level table (default working path; small, fast)
# ---------------------------------------------------------------------------
@lru_cache(maxsize=1)
def _load_brms(path: str):
    import pandas as pd
    df = pd.read_csv(path)
    df.columns = [c.strip().strip('"') for c in df.columns]
    # plot_key = STATECD-UNITCD-COUNTYCD-PLOT ; SDImax.mean in trees/ha
    return dict(zip(df["ID"].astype(str), df["SDImax.mean"].astype(float)))

def sdimax_for_plot(plot_key: str, brms_csv: str | None = None) -> float | None:
    """Localized max SDI (trees/ha) for an FIA plot_key 'STATE-UNIT-COUNTY-PLOT'."""
    brms_csv = brms_csv or os.environ.get("BRMS_SDIMAX_CSV", "")
    if not brms_csv or not os.path.exists(brms_csv):
        return None
    return _load_brms(brms_csv).get(str(plot_key))

# ---------------------------------------------------------------------------
# Source 1: TreeMap SDImax raster lookup (highest fidelity; per coordinate)
# ---------------------------------------------------------------------------
def sdimax_for_coord(lon: float, lat: float, raster_path: str | None = None) -> float | None:
    """Localized max SDI (trees/ha) at a coordinate from the TreeMap SDImax raster.
    raster_path -> the SDImax band of the TreeMap product (Zenodo 19509367)."""
    raster_path = raster_path or os.environ.get("TREEMAP_SDIMAX_TIF", "")
    if not raster_path or not os.path.exists(raster_path):
        return None
    try:
        import rasterio
        from rasterio.warp import transform as _t
        with rasterio.open(raster_path) as src:
            xs, ys = _t("EPSG:4326", src.crs, [lon], [lat])
            val = next(src.sample([(xs[0], ys[0])]))[0]
            return float(val) if val and val > 0 else None
    except Exception:
        return None

# ---------------------------------------------------------------------------
# FVS plug-in: per-stand SDIMAX keyword block
# ---------------------------------------------------------------------------
def fvs_sdimax_keywords(max_sdi_tph: float, maxsp: int, to_acre: bool = True) -> str:
    """SDIMAX keyword block setting every species' max SDI to the localized stand value,
    so FVS's basal-area-weighted stand maximum equals the localized value regardless of
    composition. to_acre converts trees/ha -> trees/acre (FVS internal units)."""
    val = max_sdi_tph / TPA_PER_HA if to_acre else max_sdi_tph
    return "\n".join("SDIMAX  %10d%10.1f" % (i, val) for i in range(1, maxsp + 1))

def localized_sdimax_keywords(plot_key=None, lon=None, lat=None, maxsp=120,
                              brms_csv=None, raster_path=None) -> str | None:
    """Resolve the localized max SDI (coordinate raster preferred, then plot table) and
    return the FVS SDIMAX keyword block, or None if no source resolves."""
    v = None
    if lon is not None and lat is not None:
        v = sdimax_for_coord(lon, lat, raster_path)
    if v is None and plot_key is not None:
        v = sdimax_for_plot(plot_key, brms_csv)
    return fvs_sdimax_keywords(v, maxsp) if v else None

File: docs/archive/conus-sf-integration/sdimax/localized_sdimax.R
## Localized maximum SDI for CONUS growth-and-yield models (R companion)
##
## Replaces the species-weighted maximum SDI (biased ~+28%, near-zero plot-level skill)
## with a per-stand, data-derived maximum from an FIA-based surface. Returns a per-stand
## maximum SDI (trees/ha) any growth-and-yield engine can use as its density limit, and
## emits the FVS SDIMAX keyword block. Model-agnostic.
##
## Source fidelity order: (1) TreeMap 2022 30 m SDImax raster (Zenodo 10.5281/zenodo.19509367),
## (2) brms FIA plot table, (3) forest-type + geography fallback.

TPA_PER_HA <- 2.4710538  # trees/ha / this = trees/acre (FVS internal units)

## brms_csv cols: STATECD,UNITCD,COUNTYCD,PLOT,ID(plot_key),SDImax.mean,SDImax.median
load_brms_sdimax <- function(brms_csv) {
  b <- utils::read.csv(brms_csv, stringsAsFactors = FALSE)
  names(b) <- trimws(gsub('"', "", names(b)))
  stats::setNames(as.numeric(b$SDImax.mean), as.character(b$ID))
}

sdimax_for_plot <- function(plot_key, brms_lookup) {
  v <- brms_lookup[[as.character(plot_key)]]
  if (is.null(v) || !is.finite(v)) NA_real_ else v
}

## TreeMap SDImax raster lookup (needs terra); raster_path -> SDImax band
sdimax_for_coord <- function(lon, lat, raster_path) {
  if (!requireNamespace("terra", quietly = TRUE) || !file.exists(raster_path)) return(NA_real_)
  r  <- terra::rast(raster_path)
  pt <- terra::vect(data.frame(lon = lon, lat = lat), geom = c("lon", "lat"), crs = "EPSG:4326")
  pt <- terra::project(pt, terra::crs(r))
  v  <- terra::extract(r, pt)[, 2]
  ifelse(is.finite(v) & v > 0, v, NA_real_)
}

## Per-stand SDIMAX keyword block: every species set to the localized stand value, so the
## BA-weighted stand maximum equals the localized value regardless of composition.
fvs_sdimax_keywords <- function(max_sdi_tph, maxsp, to_acre = TRUE) {
  val <- if (to_acre) max_sdi_tph / TPA_PER_HA else max_sdi_tph
  paste(sprintf("SDIMAX  %10d%10.1f", seq_len(maxsp), val), collapse = "\n")
}

## One-call resolver: raster preferred, then plot table; NA if nothing resolves.
localized_sdimax_keywords <- function(plot_key = NULL, lon = NULL, lat = NULL, maxsp = 120,
                                      brms_lookup = NULL, raster_path = NULL) {
  v <- NA_real_
  if (!is.null(lon) && !is.null(lat) && !is.null(raster_path)) v <- sdimax_for_coord(lon, lat, raster_path)
  if (!is.finite(v) && !is.null(plot_key) && !is.null(brms_lookup)) v <- sdimax_for_plot(plot_key, brms_lookup)
  if (!is.finite(v)) return(NA_character_)
  fvs_sdimax_keywords(v, maxsp)
}

## Non-circular validation: deviance explained of OBSERVED self-thinning by relative density
## under each max-SDI candidate. The better maximum is the one whose RD predicts observed
## density loss best. Needs mgcv.
selfthin_skill <- function(df, sdi_col = "SDI1",
                           maxsd_cols = c(brms = "SDImax_brms", fvs = "fvs_sdimax"),
                           tph1 = "TPH1", tph2 = "TPH2", years = "YEARS") {
  stopifnot(requireNamespace("mgcv", quietly = TRUE))
  df$mort <- -log(df[[tph2]] / df[[tph1]]) / df[[years]]
  df <- df[is.finite(df$mort), ]
  sapply(maxsd_cols, function(mc) {
    rd <- df[[sdi_col]] / df[[mc]]; ok <- is.finite(rd) & is.finite(df$mort)
    summary(mgcv::bam(df$mort[ok] ~ s(rd[ok], k = 8)))$dev.expl
  })
}
